[CH-148620] Implement IconButton

j-x-han opened the pull request:
Adds `IconButton` and its variants. [The ticket](https://app.clubhouse.io/sci-design-system/story/148620/implement-large-secondary-icon-button-c) originally just asked for the Large Secondary variant, but the styles weren't super different so I went ahead and added the other variants as well!

~~Similar to `Chip`, MUI uses the sizes "small" and "medium" rather than "small" and "large," so I went with "medium" to keep it consistent.~~ As per our discussions, I used `sdsType` for `primary`/`secondary` and `sdsSize` for `small`/`large`.

**Design link:** https://www.figma.com/file/EaRifXLFs54XTjO1Mlszkg/Science-Design-System-Reference?node-id=642%3A4470 

I will add the Filter and Info icon svgs in a follow up PR after Erin's change for loading svgs (https://github.com/chanzuckerberg/sci-components/pull/42)

**Large Primary:**
![ezgif com-gif-maker (15)](https://user-images.githubusercontent.com/53838890/129266711-f1dc28af-087c-48a5-ae0b-5d14b35997bb.gif)

**Large Primary Disabled:**
![Screen Shot 2021-08-12 at 1 24 06 PM](https://user-images.githubusercontent.com/53838890/129266589-adebd867-25c3-4e1f-9b81-3636291b048b.png)

**Large Secondary:**
![ezgif com-gif-maker (16)](https://user-images.githubusercontent.com/53838890/129266925-a0879739-3f83-4cea-beec-00bf6f579164.gif)

**Large Secondary Disabled:**
![Screen Shot 2021-08-12 at 1 22 33 PM](https://user-images.githubusercontent.com/53838890/129266571-e15c2e75-0173-4abd-9b01-dd96db5b825f.png)

**Small Primary:**
![ezgif com-gif-maker (17)](https://user-images.githubusercontent.com/53838890/129266940-f0964a0d-bd44-416c-9db9-2fa013da42f3.gif)

**Small Primary Disabled:**
![Screen Shot 2021-08-12 at 1 24 19 PM](https://user-images.githubusercontent.com/53838890/129266605-d3d5d563-f7b9-4a94-bbf6-ed10cdff1403.png)

**Small Secondary:**
![ezgif com-gif-maker (18)](https://user-images.githubusercontent.com/53838890/129266950-4ace9fa5-7e79-46c2-9fb6-99d87e20eb6b.gif)

**Small Secondary Disabled:**
![Screen Shot 2021-08-12 at 1 24 26 PM](https://user-images.githubusercontent.com/53838890/129266621-abeebcfc-bbdc-4d85-9d7d-013d124626de.png)

j-x-han commented on src/core/IconButton/style.ts:
```diff
@@ -0,0 +1,107 @@
+
+export interface ExtraProps extends Props {
+  active?: boolean;
+  color?: "primary" | "secondary";
```
The Figma documentation currently uses `Type` for this, but MUI has the `color` option, what do you all think?
cc @tihuan @happyimadesignr @Janeece

tihuan replied:
Aha! I think this is another use case of what we talked at the SDS sync meeting just now 😄 

Essentially, `primary` and `secondary` describe sets of component props that SDS want to apply on the components. Such as:

"primary" -- applies color: "primary", size="medium"
"secondary" -- applies color: "default", size="small"

And in [this section](https://docs.google.com/document/d/1rFBfWWzmCiLM2HAsnjM0HIBxsB31xT7hCnYGaquOiUU/edit#bookmark=id.m7r0okvy4oxp), we are now brainstorming what name we should call the prop for "primary"/"secondary" for. E.g., phenotype, genotype, preset, etc..

Please add your ideas there too 🤩 🙏 

To unblock this PR, maybe we need @Janeece @happyimadesignr to help us clarify the difference between SDS "primary" and "secondary" here. 

E.g., if it's just whether the color is "primary" (blue) or "default" (gray), or does it include other component properties like size, variant, etc..

If it's just color, then maybe it's easier to update Figma to use color="primary"/"default" to describe using primary theme color and default for a gray?

If it's more than just describing color value, then we'll need a placeholder name for the prop for now, before we finalize a standard name in the doc 😆 Maybe.. `tbd="primary"/"secondary"` lol? Or something better!

ehoops replied:
Not to rush us, but I think not having naming settled is starting to block us.  I think we should get a resolution as soon as possible.

tihuan replied:
Yeah it's used so much in design! @j-x-han @ehoops do you have the bandwidth to help drive a decision on this 🙏 ? Thank you!!

j-x-han replied:
Based on our discussion, I'll update this PR to use **genotype** as the prop for primary/secondary. Does that sound right?

Janeece replied:
Sounds right to me

tihuan replied:
@ehoops @hthomas-czi Ah I'm thinking maybe it's `phenotype`, since primary/secondary here refers to presentation difference instead of functionality difference?

I remember Harley was using MenuSelect single select vs. multiple select an example for genotype/functionality difference! 

Please correct me if I'm wrong lol Thank you!

j-x-han replied:
Oh hmm I originally said `genotype` since the Figma doc used "Type" for primary/secondary and we decided to map type -> genotype, but I can see the case for `phenotype` since the differences here are visual rather than functional.

Are we okay with using `phenotype` here like Timmy suggested?

Janeece replied:
To me the difference between primary/secondary icon buttons is functional. Primary Icon buttons behave more like a typical button (default, hover, click), where as the Secondary buttons have active/inactive stages that dictate the hover state. Primary icon buttons only hold an active state in rare instances, while a dropmenu is open. 

Admittedly I'm heavily skewed by how Aspen/IDseq use them. Would love input from @happyimadesignr @hthomas-czi on their perspectives.

hthomas-czi replied:
I agree the difference between primary and second is functional @Janeece. That's why we defined the `Type` prop in the first place, otherwise it would be a `Style` prop. 

I'm not the biggest fan of the phenotype/genotype distinction. It's clever, but maybe **too** clever. It seems unnecessarily confusing to me. I'd like to propose the following terms:

`Type` -> `Function`
`Style` -> `Appearance`

I like this because it literally describes how we've been using those props. Thoughts?

tihuan replied:
ahh got it! So it is functional difference! Sorry for missing that 😆 

For the names, I'm still thinking `sdsStyle` and `sdsType` are short and sweet, and easier to type in code 🤔

j-x-han replied:
I think `Function` could potentially run into a lot of confusion/issues since a "function" means its own thing in coding 😅 And we could potentially be passing functions into the MUI components...

Using `sdsStyle`/`sdsType` seems like a straightforward change that doesn't require relearning what each property means, so I'm in favor of using them! What do the designers think?

j-x-han commented on src/core/IconButton/index.stories.tsx:
```diff
@@ -0,0 +1,60 @@
+import DehazeIcon from "@material-ui/icons/Dehaze";
```
As mentioned in the PR description, I'll add the Filter/Info icon svgs in a followup PR, so using an MUI icon for now!

j-x-han commented on src/core/IconButton/style.ts:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    .MuiSvgIcon-root {
+      height: ${spacings?.l}px;
+      width: ${spacings?.l}px;
```
The theme doesn't seem to have a (icon) "sizes" setting, so I used the spacing variable instead, but not sure if this is too unintuitive?

tihuan replied:
Ooh this is a great point! @Janeece @happyimadesignr maybe we should define dimensions for the sizes in theme? Is it always gonna follow `spacings.l` and `spacings.s` for example?

Janeece replied:
@tihuan @j-x-han definitely agree on defining icon sizes in the theme if we arent already -- we actually already have these sizes defined at the top of  the [Icons/Illustrations page](https://www.figma.com/proto/EaRifXLFs54XTjO1Mlszkg/Science-Design-System-Reference?node-id=642%3A513&scaling=min-zoom&page-id=642%3A0) (sizing names will probably have to be updated to fit our naming standards, I'm not sure where we are with that. If IconSize fits we could go with that). 

We've defined 4 sizes that correspond to the small or large (visual) style icons. Small-style icons can be icon-s or icon-xs, Large-style icons can be icon-l or icon-lx. 

Not sure if we want to implement this as a part of this ticket or if I should make a new one to implement icon sizes?

tihuan replied:
Ooh fantastic! Thanks so much for the link, @Janeece !

Ah @j-x-han added a note in the PR description that we used medium instead of large, because MUI only has `small` and `medium` for the prop `size`. But since we actually have 4 sizes, what about `sdsSize="xs . s,  l,  xl` to differentiate from MUI's `size` props? 

Which also means... technically we could also prefix SDS's usage of `style` to be `sdsStyle` instead of using `phenotype` 😆  Am I opening a can of worms here lol??

CC: @ehoops @happyimadesignr @hthomas-czi 😂

j-x-han replied:
Thanks for the suggestions + link, I've added the four `iconSizes` to the theme!

I believe we only have 2 sizes for `IconButton` specifically (`small`/`large` in Figma, currently `small`/`medium` in this PR to match with MUI and our `Chip` component. These actually correspond to sizes `icon-s` and `icon-xl`).

Do we anticipate adding the other sizes for `IconButton`? If not, then I don't think it makes sense to change the `IconButton` prop to have all four sizes as options.

I am open to renaming the prop to `sdsStyle` or something similar, though! That could potentially help with the Figma small/large -> MUI small/medium discrepancy we're running into with some of these components.

j-x-han replied:
If there are no other objections/questions, I'll leave this as using `size?: "small" | "medium"` to stay consistent with MUI and our Chip component!

j-x-han commented on src/core/IconButton/style.ts:
```diff
@@ -0,0 +1,109 @@
+    padding: ${color === "primary" ? spacings?.xxs : 0}px;
+
+    .MuiSvgIcon-root {
+      height: 32px;
+      width: 32px;
```
Similar to above, we don't seem to have this size/spacing set anywhere in the theme, not sure what the best place to add it would be

tihuan commented:
Woohoo! Love all the GIF!! Super useful 🔥 🎉 Haven't reviewed yet, but just left a comment 😄 Thank you!

tihuan commented on src/core/IconButton/style.ts:
```diff
@@ -0,0 +1,109 @@
+  ${(props: ExtraProps) => {
+    const { active, color, disabled, size } = props;
+
+    return css`
```
niiice looks really clear 🔥 🚀 !! GO JULIE!!

Janeece commented:
@j-x-han So exciting to see all of the other variants of this component also! 
A couple of comments:
- In the SDS reference, we are calling the ones you've labeled "Medium"  as "Large" instead. Was there any input that changed that? if not, can we call it Large instead?
- For the secondary buttons (both sizes), it looks like the active state color is primary-600. If thats the case, can we updated it to primary-400 when active? And on hover over the active button the color changes to primary-600.

j-x-han commented:
> @j-x-han So exciting to see all of the other variants of this component also!
> A couple of comments:
> 
> * In the SDS reference, we are calling the ones you've labeled "Medium"  as "Large" instead. Was there any input that changed that? if not, can we call it Large instead?
> * For the secondary buttons (both sizes), it looks like the active state color is primary-600. If thats the case, can we updated it to primary-400 when active? And on hover over the active button the color changes to primary-600.

Thanks for the feedback, @Janeece !

* I'm currently using "Medium" instead of "Large" since MUI's `IconButton` only has sizes `small` and `medium`. This is similar to what I did for the `Chip` component in a previous PR. I'll defer/update the PR to whatever we decide in the above comment thread regarding IconSizes.
* I've updated the active state for Secondary buttons to primary-400!

happyimadesignr commented:
Hey sorry for the delay. I do like the sds prefix because it’s more
scalable when there are other names that we need to also make sds specific.

I’d prefer small and large, if possible!! I’m a bit confused why MUI uses
small and medium, I makes me think there will be another size called large.
It sounds like we can use small and large if we do sdsSize?

I thought we said we could keep the name tag by using an alias right? 😊

Thanks Julie!!!

On Wed, Aug 18, 2021 at 4:54 PM Julie Han ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/core/IconButton/style.ts
> <https://github.com/chanzuckerberg/sci-components/pull/43#discussion_r691679801>
> :
>
> > +      height: ${spacings?.l}px;
> +      width: ${spacings?.l}px;
>
> If there are no other objections/questions, I'll leave this as using size?:
> "small" | "medium" to stay consistent with MUI and our Chip component!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chanzuckerberg/sci-components/pull/43#discussion_r691679801>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7N5O4TQSW7I6EJAEU3YA3T5RB35ANCNFSM5B74SB7Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>
-- 

*Jennifer Tang*
Product Designer

c: +1 408-540-8858 <+14085408858>
***@***.***
http://chanzuckerberg.com <http://chanzuckerberg.com/?email>
[image: Facebook] <https://www.fb.me/chanzuckerberginitiative/>   [image:
Instagram]
<https://www.instagram.com/chanzuckerberginitiative/>

j-x-han commented:
I've updated the PR to use `sdsType` for `primary`/`secondary` and `sdsSize` for `small`/`large` as per our discussions! Let me know if there's any more confusion around these props.

tihuan reviewed (approved):
LGTM!! Thanks so much for your patience and flexibility throughout the process, @j-x-han 🙏 You're amazing!
